File: solutions/java/src/main/java/com/epi/IntersectSortedArrays2.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class IntersectSortedArrays2 {
  // @include
  public static List<Integer> intersectTwoSortedArrays(List<Integer> A,
                                                       List<Integer> B) {
    List<Integer> intersectionAB = new ArrayList<>();
    for (int i = 0; i < A.size(); ++i) {
      if ((i == 0 || A.get(i) != A.get(i - 1))
          && Collections.binarySearch(B, A.get(i)) >= 0) {
        intersectionAB.add(A.get(i));
      }
    }
    return intersectionAB;
  }
  // @exclude
}

File: solutions/python/intersect_sorted_arrays2.py
# Intersect_sorted_arrays2.h 4a4c5f91493e5e482eaa79892816c1ccefa084f4
import bisect

# @include
def binary_search(A, k):
    i = bisect.bisect_left(A, k)
    return i < len(A) and A[i] == k

def intersect_two_sorted_arrays(A, B):
    intersection_A_B = []
    for i in range(len(A)):
        if (i == 0 or A[i] != A[i - 1]) and binary_search(B, A[i]):
            intersection_A_B.append(A[i])
    return intersection_A_B
# @exclude
